_source/_posts/2019-11-19-secure-kafka-java.md

```diff
@@ -0,0 +1,458 @@
+        configurations.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        return configurations;
+    }
+
+    @Bean
+    public KafkaTemplate<String, String> kafkaTemplate() {
+        return new KafkaTemplate<>(producerFactory());
+    }
+
+}
+```
+
+This class creates a `ProducerFactory` which knows how to create producers based on the configurations you provided. You've also specified to connect to your local Kafka broker and to serialize both the key and the values with `String`.
+
+You also declared a `KafkaTemplate` bean to perform high-level operations on your producer. In other words, the template is able to do operations such as sending a message to a topic and efficiently hides under-the-hood details from you.
+
+The next step is to create the endpoint to send the message to the producer. Inside the `src/main/java/com/okta/javakafka/controller` package, create the following class:
+
+```java
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+public class KafkaController {
+
+    private KafkaTemplate<String, String> template;
+
+    public KafkaController(KafkaTemplate<String, String> template) {
+        this.template = template;
+    }
+
+    @GetMapping("/kafka/produce")
+    public void produce(@RequestParam String message) {
+        template.send("myTopic", message);
+    }
+
+}
+```
+
+As you can see, this endpoint is very simple. It injects the `KafkaTemplate` configured earlier and sends a message to `myTopic` when a `GET` request is made to `/kafka/produce`.
+
+Let's test if everything is working as expected. Run the `main` method inside the `JavaKafkaApplication` class. To run from the command line, execute the following command:
+
+```bash
+./mvnw spring-boot:run
+```
+
+Your server should be running on port 8080 and you can already make API requests against it!
+
+Go to your web browser and access <http://localhost:8080/kafka/produce?message=This is my message>.
+
+When you make a call with the command above, your application will execute the `/kafka/produce` endpoint, which sends a message to `myTopic` topic inside Kafka.
+
+But how do you know the command successfully sent a message to the topic? Right now, you don't consume messages inside your app, which means you cannot be sure!
+
+Fortunately, there is an easy way to create a consumer to test right away. Inside the `bin` folder of your Kafka directory, run the following command:
+
+```bash
+./kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic myTopic
+```
+
+Access <http://localhost:8080/kafka/produce?message=This is my message> again to see the following message in the terminal running the Kafka consumer:
```

Comment:
Should you need to send a new message to the producer? Shouldn't the first message you sent be waiting for a consumer?

Reply:
I QA'd this and the steps are accurate.